server/src/main/java/org/elasticsearch/index/engine/LuceneChangesSnapshot.java

```diff
@@ -0,0 +1,213 @@
+                          long fromSeqNo, long toSeqNo, boolean requiredFullRange) throws IOException {
+        if (fromSeqNo < 0 || toSeqNo < 0 || fromSeqNo > toSeqNo) {
+            throw new IllegalArgumentException("Invalid range; from_seqno [" + fromSeqNo + "], to_seqno [" + toSeqNo + "]");
+        }
+        this.mapperService = mapperService;
+        this.fromSeqNo = fromSeqNo;
+        this.toSeqNo = toSeqNo;
+        this.lastSeenSeqNo = fromSeqNo - 1;
+        this.requiredFullRange = requiredFullRange;
+        boolean success = false;
+        final Engine.Searcher engineSearcher = searcherFactory.get();
+        try {
+            this.searcher = new IndexSearcher(Lucene.wrapAllDocsLive(engineSearcher.getDirectoryReader()));
+            this.searcher.setQueryCache(null);
+            this.topDocs = searchOperations(searcher);
+            success = true;
+            this.onClose = engineSearcher;
+        } finally {
+            if (success == false) {
+                IOUtils.close(engineSearcher);
+            }
+        }
+    }
+
+    @Override
+    public void close() throws IOException {
+        onClose.close();
+    }
+
+    @Override
+    public int totalOperations() {
+        return Math.toIntExact(topDocs.totalHits);
+    }
+
+    @Override
+    public int overriddenOperations() {
+        return skippedOperations;
+    }
+
+    @Override
+    public Translog.Operation next() throws IOException {
+        final Translog.Operation op = nextOp();
+        if (requiredFullRange && lastSeenSeqNo < toSeqNo) {
+            final long expectedSeqNo = lastSeenSeqNo + 1;
+            if (op == null || op.seqNo() != expectedSeqNo) {
+                throw new IllegalStateException("Not all operations between min_seqno [" + fromSeqNo + "] " +
+                    "and max_seqno [" + toSeqNo + "] found; expected seqno [" + expectedSeqNo + "]; found [" + op + "]");
+            }
+        }
+        if (op != null) {
+            lastSeenSeqNo = op.seqNo();
+        }
+        return op;
+    }
+
+    private Translog.Operation nextOp() throws IOException {
+        final ScoreDoc[] scoreDocs = topDocs.scoreDocs;
+        for (; docIndex < scoreDocs.length; docIndex++) {
+            if (scoreDocs[docIndex].doc == DocIdSetIterator.NO_MORE_DOCS) {
+                return null;
+            }
+            final Translog.Operation op = readDocAsOp(scoreDocs[docIndex].doc);
+            if (op != null) {
+                return op;
+            }
+        }
+        return null;
+    }
+
+    private TopDocs searchOperations(IndexSearcher searcher) throws IOException {
+        final Query rangeQuery = LongPoint.newRangeQuery(SeqNoFieldMapper.NAME, fromSeqNo, toSeqNo);
+        final Sort sortedBySeqNoThenByTerm = new Sort(
+            new SortedNumericSortField(SeqNoFieldMapper.NAME, SortField.Type.LONG),
+            new SortedNumericSortField(SeqNoFieldMapper.PRIMARY_TERM_NAME, SortField.Type.LONG, true)
```

Comment:
As discussed - this should be needed in the future. Maybe we should remove it and instead assert that we never have duplicate seq#

Reply:
I think I miss something here because I think we need it for now but not in the future after we have a Lucene rollback. I will reach out to discuss this.

Reply:
We have dedup in this PR already (line 161-163). The `lastSeenSeqNo` is used for dedup and range check. I am fine to remove the primary sort and dedup mechanism.

Reply:
I see. I missed it. I think it's surprising to put it in `readDocAsOp` and shortcut. I'd prefer to do it in `next` where do all our state updates and then everything together. it's rare anyway and doesn't require optimization imo. That said, it's all nits. If you prefer it otherwise I'm good. Thanks for clarifying.

Reply:
I agree, we should not mutate anything in `readDocAsOp`. I will update this.

Reply:
@bleskes I moved this to `next` but we also need to dudup for nested docs then I moved this to  `readDocAsOp` again. I think we should optimize for nested docs. I am open to suggestions here.